en/controllers/middleware.rst

```diff
@@ -0,0 +1,252 @@
+Middleware
+##########
+
+Middleware give you the ability to 'wrap' your application in re-usable,
+composable layers of Request handling, or response building logic. Middleware
+are part of the new HTTP stack in CakePHP that leverages the PSR7 request and
+response interfaces. By leveraging the PSR7 standard you can use any PSR7
+compatible middleware available on `The Packagist <https://thepackagist.org>`__.
+
+CakePHP provides several middleware out of the box:
+
+* ``Cake\Error\Middleware\ErrorHandlerMiddleware`` traps exceptions from the
+  wrapped middleware and renders an error page using the
+  :doc:`/development/errors` Exception handler.
+* ``Cake\Routing\AssetMiddleware`` checks whether the request is referring to a theme
+  or plugin asset file, such as a CSS, JavaScript or image file stored in either a
+  plugin's webroot folder or the corresponding one for a Theme.
+* ``Cake\Routing\Middleware\RoutingMiddleware`` uses the ``Router`` to parse the
+  incoming URL and assign routing parameters to the request.
+* ``Cake\I18n\Middleware\LocaleSelectorMiddleware`` enables automatic language
+  switching from the ``Accept-Language`` header sent by the browser.
+
+Using Middleware
+================
+
+You attach middleware in your ``App\Application`` class' ``middleware`` method.
+If you don't have an ``App\Application`` class, see the section on
+:ref:`adding-http-stack` for more information. Your application's ``middleware``
+hook method will be called early in the request process, you can use the
+``Middleware`` object to attach middleware::
+
+    namespace App;
+
+    use Cake\Http\BaseApplication;
+
+    class Application extends BaseApplication
+    {
+        public functin middleware($middleware)
+        {
+            $error = new \Cake\Error\Middleware\ErrorHandlerMiddleware();
+            $middleware->push($error);
+            return $middleware;
+        }
+    }
+
+In addition to pushing onto the end of the ``MiddlewareStack`` you can do
+a variety of operations::
+
+        $layer = new \App\Middleware\CustomMiddleware;
+
+        // Pushed middleware will be last in line.
+        $middleware->push($layer);
+
+        // Prepended middleware will be first in line.
+        $middleware->prepend($layer);
+
+        // Insert in a specific slot. If the slot is out of
+        // bounds, it will be added to the end of the line.
+        $middleware->insertAt(2, $layer);
+
+        // Insert before another middleware.
+        // If the named class cannot be found the new layer
+        // we be added to the end of the line.
```

Comment:
Yes. If you expect this you can still try/catch. Silently doing this could be dark magic.